src/server.c

```diff
@@ -17,1657 +17,609 @@
-{
-    /* Interrupt the proactor, async-signal-safe */
-    pn_proactor_interrupt(qd->server->proactor);
-}
+        if (pn_conn) {
+            qd_connection_t* conn = (qd_connection_t*) pn_connection_get_context(pn_conn);
 
-void qd_server_activate(qd_connection_t *ctx)
-{
-    if (ctx) ctx->wake(ctx);
-}
+            if (conn) {
+                qd_conn_event_batch_complete(server->container, conn, false);
+            }
+        }
 
+        pn_proactor_done(proactor, batch);
+    }
 
-void qd_connection_set_context(qd_connection_t *conn, void *context)
-{
-    conn->user_context = context;
+    return NULL;
 }
 
+// Wake function for proactor-managed connections
+static void connection_wake(qd_connection_t* conn) {
+    assert(conn);
 
-void *qd_connection_get_context(qd_connection_t *conn)
-{
-    return conn->user_context;
+    pn_connection_wake(conn->pn_conn);
 }
 
-
-void *qd_connection_get_config_context(qd_connection_t *conn)
-{
-    return conn->context;
+static void data_put_symbol(pn_data_t* data, const char* value) {
+    pn_data_put_symbol(data, pn_bytes(strlen(value), value));
 }
 
+static void data_put_string(pn_data_t* data, const char* value) {
+    pn_data_put_string(data, pn_bytes(strlen(value), value));
+}
 
-void qd_connection_set_link_context(qd_connection_t *conn, void *context)
-{
-    conn->link_context = context;
+static void data_put_map_entry(pn_data_t* data, const char* key, const char* value) {
+    data_put_symbol(data, key);
+    data_put_string(data, value);
 }
 
+static void connection_decorate(qd_server_t* server, qd_connection_t* conn, const qd_server_config_t* config) {
+    pn_connection_t* pn_conn = conn->pn_conn;
 
-void *qd_connection_get_link_context(qd_connection_t *conn)
-{
-    return conn->link_context;
-}
+    // Set the container name
+    pn_connection_set_container(pn_conn, server->container_name);
 
+    // Offer ANONYMOUS_RELAY capability
+    size_t clen = strlen(QD_CAPABILITY_ANONYMOUS_RELAY);
+    pn_data_put_symbol(pn_connection_offered_capabilities(pn_conn),
+                       pn_bytes(clen, (char*) QD_CAPABILITY_ANONYMOUS_RELAY));
 
-pn_connection_t *qd_connection_pn(qd_connection_t *conn)
-{
-    return conn->pn_conn;
-}
+    // Create the connection properties map
 
+    pn_data_t* props = pn_connection_properties(pn_conn);
 
-bool qd_connection_inbound(qd_connection_t *conn)
-{
-    return conn->listener != 0;
-}
+    pn_data_put_map(props);
+    pn_data_enter(props);
 
+    data_put_map_entry(props, QD_CONNECTION_PROPERTY_PRODUCT_KEY, QD_CONNECTION_PROPERTY_PRODUCT_VALUE);
+    data_put_map_entry(props, QD_CONNECTION_PROPERTY_VERSION_KEY, QPID_DISPATCH_VERSION);
 
-uint64_t qd_connection_connection_id(qd_connection_t *conn)
-{
-    return conn->connection_id;
-}
+    data_put_symbol(props, QD_CONNECTION_PROPERTY_CONN_ID);
+    pn_data_put_int(props, conn->connection_id);
 
+    // XXX When is config going to be null?
+    if (config && config->inter_router_cost > 1) {
```

Comment:
Btw, this looks to me like a response to a Coverity warning. Coverity likes to sends reports about NULL config ;P The Coverity results on the page usually have something they call "static trace", which is trying to answer this question...